mjpc/tasks/humanoid/tracking/task.cc

```diff
@@ -0,0 +1,231 @@
+    std::string linvel_sensor_name = "tracking_linvel[" + body_name + "]";
+    int mocap_body_id = mj_name2id(model, mjOBJ_BODY, mocap_body_name.c_str());
+    assert(0 <= mocap_body_id);
+    int body_mocapid = model->body_mocapid[mocap_body_id];
+    assert(0 <= body_mocapid);
+
+    // compute finite-difference velocity
+    mju_copy3(
+      &residual[counter],
+      model->key_mpos + model->nmocap * 3 * key_index_1 + 3 * body_mocapid);
+    mju_subFrom3(
+      &residual[counter],
+      model->key_mpos + model->nmocap * 3 * key_index_0 + 3 * body_mocapid);
+    mju_scl3(&residual[counter], &residual[counter], fps);
+
+    // subtract current velocity
+    double* sensor_linvel = mjpc::SensorByName(
+        model, data, linvel_sensor_name.c_str());
+    mju_subFrom3(&residual[counter], sensor_linvel);
+
+    counter += 3;
+  }
+
+  // sensor dim sanity check
+  // TODO: use this pattern everywhere and make this a utility function
+  int user_sensor_dim = 0;
+  for (int i=0; i < model->nsensor; i++) {
+    if (model->sensor_type[i] == mjSENS_USER) {
+      user_sensor_dim += model->sensor_dim[i];
+    }
+  }
+  if (user_sensor_dim != counter) {
+    std::printf("user_sensor_dim=%d, counter=%d", user_sensor_dim, counter);
+    mju_error_i("mismatch between total user-sensor dimension "
+                "and actual length of residual %d", counter);
+  }
+
+}
+
+// --------------------- Transition for humanoid task -------------------------
+//   Set `data->mocap_pos` based on `data->time` to move the mocap sites.
+//   Linearly interpolate between two consecutive key frames in order to
+//   smooth the transitions between keyframes.
+// ----------------------------------------------------------------------------
+int humanoid::Tracking::Transition(int state, const mjModel* model,
+                                   mjData* data, Task* task) {
+  // Hardcoded constant matching keyframes from CMU mocap dataset.
+  float fps = 30.0;
+  double current_index = data->time * fps;
+  int last_key_index = (model->nkey) - 1;
+
+  // Positions:
+  // We interpolate linearly between two consecutive key frames in order to
+  // provide smoother signal for tracking.
+  int key_index_0, key_index_1;
+  double weight_0, weight_1;
+  std::tie(key_index_0, key_index_1, weight_0, weight_1) =
+    compute_interpolation_values(current_index, last_key_index);
+
+  double mocap_pos_0[3 * model->nmocap];
+  double mocap_pos_1[3 * model->nmocap];
+
+  // Compute interpolated frame.
+  mju_scl(mocap_pos_0,
+          model->key_mpos + model->nmocap * 3 * key_index_0,
+          weight_0,
+          model->nmocap * 3);
+
+  mju_scl(mocap_pos_1,
+          model->key_mpos + model->nmocap * 3 * key_index_1,
+          weight_1,
+          model->nmocap * 3);
+
+  mju_copy(data->mocap_pos, mocap_pos_0, model->nmocap * 3);
+  mju_addTo(data->mocap_pos, mocap_pos_1, model->nmocap * 3);
+
+  int new_state = key_index_0;
+
+  return new_state;
```

Comment:
I guess here's the thing I don't understand, why do we still need to return a new state for every step? Basically, I think the states should correspond to the different clips (this can happen in a followup), but even now, is the return value used at all anywhere?

Basically, I want to move away from micro-states, now that we can use time properly. This state (which should be renamed "stage" everywhere, to avoid confusion, also out of scope), should be named things with buttons, e.g. "monkey flip", etc. Part of this change would involve using userdata to save stuff by the transition. So when one clip ends, I save in userdata the current time and the current xy position of the root, and then in the residual that gives me the spatio-temporal origin w.r.t which I shift my clip data. Does that make sense?
@thowell, you agree? Basically we should treat userdata like we do mocap everywhere, it's a thing that gets copied in as an initial state.

Anyway that was a bit of a rant. Point being, can we `return 0` or will something break?

Reply:
No need to return `new_state`, `return 0` is fine. @yuvaltassa we should think about changing the `Transition` interface. Probably no reason to return a "state" since it's a member attribute and can be updated internally. 

My suggestion for now is to only support a single motion tracking example. Then, in a follow-up PR we should add support for switching between motions (or continuously running through all of them) via options exposed in the GUI.

@yuvaltassa we should copy userdata and as you suggest and consider it part of the full state.

Reply:
`return 0` is fine here. I just pushed a change.